In [0]:
from __future__ import print_function
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
import pandas as pd

In [0]:
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()


        self.conv1 = nn.Sequential(
            nn.Conv2d(in_channels=1, out_channels=32, kernel_size=(3,3), padding=0, stride=1), 
            nn.ReLU()
        )#Output=26 RF=3X3 [RFin + (Ksize-1 * JMPin) => 1+(3-1)*1 =3]  :JMPin=1, Jout= JMPin X s = 1
        self.conv2 = nn.Sequential(
            nn.Conv2d(in_channels=32, out_channels=64, kernel_size=(3,3), padding=0, stride=1),
            nn.ReLU()    
        )#Output=24 RF=5X5  [RFin + (Ksize-1 * JMPin) => 3+(3-1)*1 =5] :JMPin=1, Jout =JMPin X s =1

        self.pool1 = nn.MaxPool2d(2, 2)#Output=12 RF=6X6 [RFin + (Ksize-1 * JMPin) => 5+(2-1)*1 =6] :JMPin=1, Jout=  JMPin X s =2

        self.conv3 = nn.Sequential(
            nn.Conv2d(in_channels=64, out_channels=128, kernel_size=(3,3), padding=0, stride=1),
            nn.ReLU()
        )#Output=10 RF=10X10 [RFin + (Ksize-1 * JMPin) => 6+(3-1)*2 =10] : Jout= JMPin X s = 2X1 :JMPin=2, Jout= JMPin X s = 2X1=2

        self.pool2 = nn.MaxPool2d(2, 2) #Output=5 RF=12[RFin + (Ksize-1 * JMPin) => 10+(2-1)*2 =12]  :JMPin=2, Jout =JMPin X s = 2X2 =4

        self.conv4 = nn.Sequential(
            nn.Conv2d(in_channels=128, out_channels=256, kernel_size=(3,3), padding=0, stride=1),
            nn.ReLU()            
        )#Output=3 RF= 20[RFin + (Ksize-1 * JMPin) => 12+(3-1)*4 =20] :JMPin=4, Jout =JMPin X s = 4X1=4

        
        self.conv5 = nn.Sequential(
            nn.Conv2d(in_channels=256, out_channels=10, kernel_size=(3,3), padding=0, stride=1),
            #nn.ReLU()
        )#Output=1 RF=28 [RFin + (Ksize-1 * JMPin) => 20+(3-1)*4 =28]  :JMPin=4, Jout=JMPin X s = 4X1=4
     

    def forward(self, x):
        x = self.conv1(x)
        x = self.conv2(x)
        x = self.pool1(x)
        x = self.conv3(x)
        x = self.pool2(x)
        x = self.conv4(x)
        x = self.conv5(x)        
        x = x.view(-1, 10)
        return F.log_softmax(x)

In [4]:
!pip install torchsummary
from torchsummary import summary
use_cuda = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")
model = Net().to(device)
summary(model, input_size=(1, 28, 28))

----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1           [-1, 32, 26, 26]             320
              ReLU-2           [-1, 32, 26, 26]               0
            Conv2d-3           [-1, 64, 24, 24]          18,496
              ReLU-4           [-1, 64, 24, 24]               0
         MaxPool2d-5           [-1, 64, 12, 12]               0
            Conv2d-6          [-1, 128, 10, 10]          73,856
              ReLU-7          [-1, 128, 10, 10]               0
         MaxPool2d-8            [-1, 128, 5, 5]               0
            Conv2d-9            [-1, 256, 3, 3]         295,168
             ReLU-10            [-1, 256, 3, 3]               0
           Conv2d-11             [-1, 10, 1, 1]          23,050
Total params: 410,890
Trainable params: 410,890
Non-trainable params: 0
----------------------------------------------------------------
Input size (MB): 0.00
Forward/

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:45: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.


In [5]:


torch.manual_seed(1)
batch_size = 128

kwargs = {'num_workers': 1, 'pin_memory': True} if use_cuda else {}
train_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=True, download=True,
                    transform=transforms.Compose([
                        transforms.ToTensor(),
                        transforms.Normalize((0.1307,), (0.3081,))
                    ])),
    batch_size=batch_size, shuffle=True, **kwargs)
test_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=False, transform=transforms.Compose([
                        transforms.ToTensor(),
                        transforms.Normalize((0.1307,), (0.3081,))
                    ])),
    batch_size=batch_size, shuffle=True, **kwargs)


Extracting ../data/MNIST/raw/train-images-idx3-ubyte.gz to ../data/MNIST/raw


Extracting ../data/MNIST/raw/train-labels-idx1-ubyte.gz to ../data/MNIST/raw


Extracting ../data/MNIST/raw/t10k-images-idx3-ubyte.gz to ../data/MNIST/raw


Extracting ../data/MNIST/raw/t10k-labels-idx1-ubyte.gz to ../data/MNIST/raw
Processing...
Done!





In [0]:
from tqdm import tqdm

Train_Loss=[]
Train_Accuracy=[]
Test_Loss=[]
Test_Accuracy=[]

def train(model, device, train_loader, optimizer, epoch):
    model.train()
    pbar = tqdm(train_loader)
    correct_train=0
    for batch_idx, (data, target) in enumerate(pbar):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        Train_Loss.append(loss.item())
        pred = output.argmax(dim=1, keepdim=True)
        correct_train += pred.eq(target.view_as(pred)).sum().item()
        Train_Accuracy.append(100.00 *correct_train/len(train_loader.dataset))
        loss.backward()
        optimizer.step()
        #pbar.set_description(desc= f'TRAINING Loss={loss.item()} batch_id={batch_idx}')
        
        ##Added detailed percentage:
        pbar.set_description(desc= f'TRAIN Loss={loss.item()} batch_id={batch_idx} Correct={correct_train} / {len(train_loader.dataset)} TRAIN ACCURACY={100.00 *correct_train/len(train_loader.dataset)}')
     


def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            loss=F.nll_loss(output, target)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            
            correct += pred.eq(target.view_as(pred)).sum().item()
            
    test_loss /= len(test_loader.dataset)
    Test_Loss.append(test_loss)
    Test_Accuracy.append(100. * correct / len(test_loader.dataset))
    print('\nTEST: Average loss: {:.4f}, Accuracy: {}/{} ({:.2f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))
    


In [8]:
import pandas as pd
import numpy as np

model = Net().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.9)
df_Test_Train_Analysis=pd.DataFrame(columns=['EPOCH','Test_Acc','Train_Acc','Acc_Diff'])
print(df_Test_Train_Analysis.shape)
for epoch in range(1, 20):
    print('EPOCH #',epoch)
    train(model, device, train_loader, optimizer, epoch)
    test(model, device, test_loader)
    print('----------------------------------------------------------------------')

  0%|          | 0/469 [00:00<?, ?it/s]

(0, 4)
EPOCH # 1


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:45: UserWarning: Implicit dimension choice for log_softmax has been deprecated. Change the call to include dim=X as an argument.
TRAIN Loss=0.05867135897278786 batch_id=468 Correct=54886 / 60000 TRAIN ACCURACY=91.47666666666667: 100%|██████████| 469/469 [00:12<00:00, 38.09it/s]
  0%|          | 0/469 [00:00<?, ?it/s]


TEST: Average loss: 0.0610, Accuracy: 9818/10000 (98.18%)

----------------------------------------------------------------------
EPOCH # 2


TRAIN Loss=0.008618916384875774 batch_id=468 Correct=59059 / 60000 TRAIN ACCURACY=98.43166666666667: 100%|██████████| 469/469 [00:11<00:00, 39.28it/s]
  0%|          | 0/469 [00:00<?, ?it/s]


TEST: Average loss: 0.0346, Accuracy: 9884/10000 (98.84%)

----------------------------------------------------------------------
EPOCH # 3


TRAIN Loss=0.020571837201714516 batch_id=468 Correct=59299 / 60000 TRAIN ACCURACY=98.83166666666666: 100%|██████████| 469/469 [00:11<00:00, 39.85it/s]
  0%|          | 0/469 [00:00<?, ?it/s]


TEST: Average loss: 0.0309, Accuracy: 9899/10000 (98.99%)

----------------------------------------------------------------------
EPOCH # 4


TRAIN Loss=0.02268158830702305 batch_id=468 Correct=59472 / 60000 TRAIN ACCURACY=99.12: 100%|██████████| 469/469 [00:11<00:00, 39.32it/s]
  0%|          | 0/469 [00:00<?, ?it/s]


TEST: Average loss: 0.0279, Accuracy: 9911/10000 (99.11%)

----------------------------------------------------------------------
EPOCH # 5


TRAIN Loss=0.06169891729950905 batch_id=468 Correct=59570 / 60000 TRAIN ACCURACY=99.28333333333333: 100%|██████████| 469/469 [00:11<00:00, 39.98it/s]
  0%|          | 0/469 [00:00<?, ?it/s]


TEST: Average loss: 0.0305, Accuracy: 9910/10000 (99.10%)

----------------------------------------------------------------------
EPOCH # 6


TRAIN Loss=0.07291072607040405 batch_id=468 Correct=59646 / 60000 TRAIN ACCURACY=99.41: 100%|██████████| 469/469 [00:11<00:00, 39.70it/s]
  0%|          | 0/469 [00:00<?, ?it/s]


TEST: Average loss: 0.0319, Accuracy: 9879/10000 (98.79%)

----------------------------------------------------------------------
EPOCH # 7


TRAIN Loss=0.00857947301119566 batch_id=468 Correct=59697 / 60000 TRAIN ACCURACY=99.495: 100%|██████████| 469/469 [00:11<00:00, 39.68it/s]
  0%|          | 0/469 [00:00<?, ?it/s]


TEST: Average loss: 0.0272, Accuracy: 9916/10000 (99.16%)

----------------------------------------------------------------------
EPOCH # 8


TRAIN Loss=0.010616093873977661 batch_id=468 Correct=59771 / 60000 TRAIN ACCURACY=99.61833333333334: 100%|██████████| 469/469 [00:11<00:00, 39.67it/s]
  0%|          | 0/469 [00:00<?, ?it/s]


TEST: Average loss: 0.0276, Accuracy: 9923/10000 (99.23%)

----------------------------------------------------------------------
EPOCH # 9


TRAIN Loss=0.0027467410545796156 batch_id=468 Correct=59817 / 60000 TRAIN ACCURACY=99.695: 100%|██████████| 469/469 [00:11<00:00, 41.22it/s]
  0%|          | 0/469 [00:00<?, ?it/s]


TEST: Average loss: 0.0267, Accuracy: 9923/10000 (99.23%)

----------------------------------------------------------------------
EPOCH # 10


TRAIN Loss=0.026645123958587646 batch_id=468 Correct=59828 / 60000 TRAIN ACCURACY=99.71333333333334: 100%|██████████| 469/469 [00:11<00:00, 40.19it/s]
  0%|          | 0/469 [00:00<?, ?it/s]


TEST: Average loss: 0.0257, Accuracy: 9920/10000 (99.20%)

----------------------------------------------------------------------
EPOCH # 11


TRAIN Loss=0.0004986723070032895 batch_id=468 Correct=59875 / 60000 TRAIN ACCURACY=99.79166666666667: 100%|██████████| 469/469 [00:11<00:00, 39.66it/s]
  0%|          | 0/469 [00:00<?, ?it/s]


TEST: Average loss: 0.0236, Accuracy: 9932/10000 (99.32%)

----------------------------------------------------------------------
EPOCH # 12


TRAIN Loss=0.0003327826561871916 batch_id=468 Correct=59883 / 60000 TRAIN ACCURACY=99.805: 100%|██████████| 469/469 [00:11<00:00, 39.82it/s]
  0%|          | 0/469 [00:00<?, ?it/s]


TEST: Average loss: 0.0297, Accuracy: 9914/10000 (99.14%)

----------------------------------------------------------------------
EPOCH # 13


TRAIN Loss=0.0002436836512060836 batch_id=468 Correct=59932 / 60000 TRAIN ACCURACY=99.88666666666667: 100%|██████████| 469/469 [00:11<00:00, 41.42it/s]
  0%|          | 0/469 [00:00<?, ?it/s]


TEST: Average loss: 0.0273, Accuracy: 9933/10000 (99.33%)

----------------------------------------------------------------------
EPOCH # 14


TRAIN Loss=0.0078002214431762695 batch_id=468 Correct=59941 / 60000 TRAIN ACCURACY=99.90166666666667: 100%|██████████| 469/469 [00:11<00:00, 40.37it/s]
  0%|          | 0/469 [00:00<?, ?it/s]


TEST: Average loss: 0.0319, Accuracy: 9914/10000 (99.14%)

----------------------------------------------------------------------
EPOCH # 15


TRAIN Loss=0.0005946258897893131 batch_id=468 Correct=59936 / 60000 TRAIN ACCURACY=99.89333333333333: 100%|██████████| 469/469 [00:11<00:00, 41.21it/s]
  0%|          | 0/469 [00:00<?, ?it/s]


TEST: Average loss: 0.0286, Accuracy: 9930/10000 (99.30%)

----------------------------------------------------------------------
EPOCH # 16


TRAIN Loss=0.00017061829566955566 batch_id=468 Correct=59941 / 60000 TRAIN ACCURACY=99.90166666666667: 100%|██████████| 469/469 [00:11<00:00, 40.84it/s]
  0%|          | 0/469 [00:00<?, ?it/s]


TEST: Average loss: 0.0344, Accuracy: 9921/10000 (99.21%)

----------------------------------------------------------------------
EPOCH # 17


TRAIN Loss=0.007293244358152151 batch_id=468 Correct=59934 / 60000 TRAIN ACCURACY=99.89: 100%|██████████| 469/469 [00:11<00:00, 41.54it/s]
  0%|          | 0/469 [00:00<?, ?it/s]


TEST: Average loss: 0.0297, Accuracy: 9927/10000 (99.27%)

----------------------------------------------------------------------
EPOCH # 18


TRAIN Loss=5.371372026274912e-05 batch_id=468 Correct=59965 / 60000 TRAIN ACCURACY=99.94166666666666: 100%|██████████| 469/469 [00:11<00:00, 40.79it/s]
  0%|          | 0/469 [00:00<?, ?it/s]


TEST: Average loss: 0.0293, Accuracy: 9924/10000 (99.24%)

----------------------------------------------------------------------
EPOCH # 19


TRAIN Loss=0.0001220405101776123 batch_id=468 Correct=59900 / 60000 TRAIN ACCURACY=99.83333333333333: 100%|██████████| 469/469 [00:11<00:00, 39.30it/s]



TEST: Average loss: 0.0313, Accuracy: 9921/10000 (99.21%)

----------------------------------------------------------------------


#Goal	:
To create Basic skeleton of a Model with systematic Layers & Maxpooling defined.Additionally defined the skeleton of the code to incorporate various Print statemetns that would help to capture Model performance at every Epoch.
#Params	4,10,890
#Best Train Accuracy	
99.94
#Best Test Accuracy	
98.99
#Observation/ Analysis/Conclusion
The Model is overall decent Model. By that we mean, at few of the places test accuracy is GREATER THAN Train accuracy. And most of the Epochs model is Not Overfitting
#Comment	:
We can Proceed confidently to continue to improve this model.